### PR TITLE
Containerize project : configure dockerfile and compose files

### LINKED ISSUE #740: Containerize this Project for Easier Contribution
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when trying to set up the development environment for this project. The process can be time-consuming and error-prone, particularly when dealing with different dependencies and system configurations.

**Describe the solution you'd like**
I propose containerizing this project using Docker. By providing a pre-configured Docker container, developers can quickly and easily set up a consistent development environment. This would streamline the onboarding process for new contributors and ensure that everyone is working with the same setup, reducing the likelihood of environment-related issues.

### PR DESCRIPTION
**Pull request by sujal-goswami**
resolve #740 

For user to build this project using docker they need to fulfil the prerequisites:

## Prerequistes
-  Docker Desktop or docker should be installed
-  WSL - Windows Subsystem for Linux  (only for windows users)
- should have `X server` installed in host machine
if not have then can download from command `sudo apt-get install x11-xserver-utils`.
to check its correct installation run `xhost` this should give output as 
```
access control disabled, clients can connect from any host
LOCAL:
SI:localuser:wslg
```
## demo image
wave

## References
[https://github.com/cypress-io/cypress-documentation/issues/5816](https://github.com/cypress-io/cypress-documentation/issues/5816)
[--no-sandbox error](https://ask.replit.com/t/get-error-by-command-chromium-browser-no-sandbox/14514/2)

### CODE REVIEW

**Comment by esimkowitz** [comment]
Thank you for setting this up! Since this is for a dev container, I think the .dockerignore doesn't need to be as restrictive, since the developer may care about the .md, config, and Yarn files. Otherwise it looks good to me!

**Review by esimkowitz — CHANGES_REQUESTED** [review]
Based on your issue, I assumed you meant to create a docker container for development purposes, usually for something like this you would want to add the VSCode tooling for Dev Containers and leave the repo data inside the final container. (see [Developing inside a Container](https://code.visualstudio.com/docs/devcontainers/containers)

I'm not sure I see a point to just running our app in a container, since it won't have access to the rest of your system, which kind of defeats the purpose of a terminal app.

**Comment by esimkowitz** [comment]
Could you verify that this complies with the guidelines for the dev containers in the guide I linked ([Developing inside a Container](https://code.visualstudio.com/docs/devcontainers/containers))? This one may also be helpful: [Create a development container](https://code.visualstudio.com/docs/devcontainers/create-dev-container#_path-to-creating-a-dev-container). I think you can skip setting a CMD to start on container run, since this is not necessary for developing inside a container. Also I realize the suggested changes I made left a bunch of new lines, could you clean those up?

**Comment by sujal-goswami** [comment]
Sure !

**Comment by sujal-goswami** [comment]
now I think it is good to go; now, the user can use the VS Code dev container feature for seamless development in the container with pre-install go , eslint and prettier extensions in it.

**Comment by esimkowitz** [comment]
I don't think you need to build the app at all in the dockerfile, since it's fast enough. Also, because webpack-watch is blocking, the dockerfile never completes. I moved this to the "postStartCommand" in devcontainer.json, but I'm still unable to launch the app, as it's complaining about DBus not being started. I tried moving the `service dbus start` command to the postStartCommand too, but it's still not working. What host OS are you using to test this? I'm on macOS 14.

I'm concerned that if the devcontainer isn't reliable on all platforms or requires additional hoops to get working, it won't be that much better than just installing the deps directly. Aside from scripthaus, we don't have any crazy dependencies. We're also in the process of a pretty large rearchitecting and we're removing the scripthaus dependency as part of this, so our dependency story will get a lot simpler.

**Comment by sujal-goswami** [comment]
I don't think the host OS is the issue here. It's more likely an issue with the dbus service. I encountered a similar dbus error during my testing but resolved it by running `service dbus start` on the machine. You can refer to this [link](https://ask.replit.com/t/get-error-by-command-chromium-browser-no-sandbox/14514/2) for more details. By the way, I'm using Windows as the host OS, so it's odd if the OS is indeed the problem. The rearchitecting plan sounds great! Simplifying the dependencies will definitely streamline the process.

**Comment by klmcwhirter** [comment]
FWIW, I did something similar as this PR - but, my goal was to build waveterm in a container - not run it in the container.

I purposely do not install multiple tool chains, project specific dependencies, etc. on localhost.

FYI, I am running on Fedora 41 and typically use VS Code (although I have been using `emacs -nw` again lately.

I concur that .devcontainer might be of dubious value, and there are several missing extensions that I typically use for web dev (vscode-tailwindcss, etc.). So I would probably just create my own.

In my case, I chose fedora:41 as the base OCI image because I wanted to be sure the resulting binary would run on my system.

That is not an issue when running in a container. If I was looking for containerization, I would prefer flatpak as an execution env because it is a more natural fit.

I suspect there may be a number of personal preferences out there.

But agree with @esimkowitz - thanks for putting in the effort. I am sure there are devs out there who can benefit.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    apt-get upgrade -y && \
+    apt-get install -y golang-go nodejs npm && \
+    npm install -g corepack && \
+    corepack enable && \
+    yarn install && \
+    apt-get install -y libgtk2.0-0t64 libgtk-3-0t64 libgbm-dev libnotify-dev libnss3 libxss1 libasound2t64 libxtst6 xauth xvfb && \
+    export DISPLAY=:0 && \
+    service dbus start
+
+RUN git clone https://github.com/scripthaus-dev/scripthaus.git
```
Comment (klmcwhirter) [inline Dockerfile:14]: Someone probably already has commented on this ...

But is this the github project you really intended to merge? note scripthaus-dev ...

**Comment by esimkowitz** [comment]
Closing this PR for now, sorry we couldn't get it merged

**Pull request closed without merging**